pkg/runtime/pubsub/outbox.go

```diff
@@ -0,0 +1,305 @@
+	o.lock.RLock()
+	c, ok := o.outboxStores[stateStore]
+	o.lock.RUnlock()
+
+	if !ok {
+		return nil, fmt.Errorf("error publishing internal outbox message: could not find outbox configuration on state store %s", stateStore)
+	}
+
+	trs := make([]state.TransactionalStateOperation, 0, len(operations))
+	for _, op := range operations {
+		sr, ok := op.(state.SetRequest)
+		if ok {
+			tr, err := transaction()
+			if err != nil {
+				return nil, err
+			}
+
+			var ceData []byte
+			bt, ok := sr.Value.([]byte)
+			if ok {
+				ceData = bt
+			} else {
+				ceData = []byte(fmt.Sprintf("%v", sr.Value))
+			}
+
+			ce := &CloudEvent{
+				ID:     tr.GetKey(),
+				Source: source,
+				Pubsub: c.outboxPubsub,
+				Data:   ceData,
+			}
+
+			if sr.ContentType != nil {
+				ce.DataContentType = *sr.ContentType
+			}
+
+			msg, err := NewCloudEvent(ce, nil)
+			if err != nil {
+				return nil, err
+			}
+
+			data, err := json.Marshal(msg)
+			if err != nil {
+				return nil, err
+			}
+
+			err = o.publishFn(ctx, &contribPubsub.PublishRequest{
+				PubsubName: c.outboxPubsub,
+				Data:       data,
+				Topic:      outboxTopic(source, c.publishTopic),
+			})
+			if err != nil {
+				return nil, err
+			}
+
+			trs = append(trs, tr)
+		}
+	}
+
+	return trs, nil
+}
+
+func outboxTopic(appID string, topic string) string {
+	return appID + topic + "outbox"
+}
+
+func (o *outboxImpl) SubscribeToInternalTopics(ctx context.Context, appID string) error {
+	o.lock.RLock()
+	defer o.lock.RUnlock()
+
+	for stateStore, c := range o.outboxStores {
+		outboxPubsub, ok := o.getPubsubFn(c.outboxPubsub)
+		if !ok {
+			outboxLogger.Warnf("could not subscribe to internal outbox topic: outbox pubsub %s not loaded", c.outboxPubsub)
+			continue
+		}
+
+		outboxPubsub.Subscribe(ctx, contribPubsub.SubscribeRequest{
+			Topic: outboxTopic(appID, c.publishTopic),
+		}, func(ctx context.Context, msg *contribPubsub.NewMessage) error {
+			var cloudEvent map[string]interface{}
+
+			err := json.Unmarshal(msg.Data, &cloudEvent)
+			if err != nil {
+				return err
+			}
+
+			stateKey := o.cloudEventExtractorFn(cloudEvent, contribPubsub.IDField)
+			data := []byte(o.cloudEventExtractorFn(cloudEvent, contribPubsub.DataField))
+			contentType := o.cloudEventExtractorFn(cloudEvent, contribPubsub.DataContentTypeField)
+			d, err := time.ParseDuration(c.stateScanDelay)
+			if err != nil {
+				d, _ = time.ParseDuration(defaultStateScanDelay)
+			}
+
+			store, ok := o.getStateFn(stateStore)
+			if !ok {
+				return fmt.Errorf("cannot get outbox state: state store %s not found", stateStore)
+			}
+
+			time.Sleep(d)
+			policyRunner := resiliency.NewRunner[struct{}](ctx,
+				resiliency.NewPolicyDefinition(outboxLogger, "outbox", d, &retry.Config{
+					Policy:      retry.PolicyExponential,
+					MaxInterval: time.Second * 15,
+					MaxRetries:  3,
```

Comment:
the fact that this does finite retries, makes it not providing a strong guarantee that the final event will be delivered... with the limitations that exist there is little things you can do about this... one possible improvement is to make the max retries configurable. Still does not provide a guarantee of strong consistency but should help for most of the cases

Reply:
This has been changed in a follow up PR, there's now a strong guarantee.